// vue-rainbow/webpack.config.js
var path = require('path');
var webpack = require('webpack');
var relativeToRootPath = ".";
var WebpackShellPlugin = require('webpack-shell-plugin');
module.exports = {
  entry: {
    'test': ['./test/src/index.js']
  },
  output: {
    path: path.resolve(__dirname, './test/dist'),
    publicPath: './',
    filename: 'build.js'
  },
  resolveLoader: {
    root: path.join(__dirname, 'node_modules')
  },
  module: {
    preloaders: [{
      test: /\.scss/,
      loader: 'import-glob-loader'
    }],
    loaders: [
      {
        test: /\.js$/,
        loader: 'babel',
        exclude: /node_modules\/(?!@dp\/url-rewrite)/
      },
      {
        test: /\.vue$/,
        loader: 'vue'
      },
      {
        test: /\.js$/,
        loader: 'babel',
        exclude: /node_modules/
      },
      {
        test: /\.json$/,
        loader: 'json'
      },
      {
        test: /\.html$/,
        loader: 'vue-html'
      },
      {
          test: /\.less$/,
          loader: "style!css!postcss!less"
      },
      {
        test: /\.(jpe?g|png|gif|svg)$/i,
        loaders: [
          'url?limit=25000'
        ]
      }
    ]
  },
  plugins: [
  ],
  devServer: {
    historyApiFallback: false,
    noInfo: true,
    hot: true,
    contentBase: './test/html',
    publicPath: '/',
    stats:{
      colors: true
    }
  },
  devtool: '#source-map'
};
