// rainbow-enums/enums/draw-status.js
/*
  检查抽奖接口返回状态，直接对接后端data.result里面的返回状态 * 1
 */
//SUCCESS("000000","成功")
export const SUCCESS = 'SUCCESS';
//USER_NOT_LOGIN("110001","用户未登陆")
export const USER_NOT_LOGIN = 'USER_NOT_LOGIN';
//USER_STAUS_ABNORMAL("110002","很遗憾，您没有抽中~")
export const USER_STAUS_ABNORMAL = 'USER_STAUS_ABNORMAL';
//UNMATCHED_USER("110003","请重新登录"),
export const UNMATCHED_USER = 'UNMATCHED_USER';
//USER_NOT_EXIST("110004","用户不存在")
export const USER_NOT_EXIST = 'USER_NOT_EXIST';
//INPUT_INVALID_ERROR("30001", "输入参数错误")
export const INPUT_INVALID_ERROR = 'INPUT_INVALID_ERROR';
//ACTIVITY_NOT_EXIST("300002","该活动不存在")
export const ACTIVITY_NOT_EXIST = 'ACTIVITY_NOT_EXIST';
//ACTIVITY_NOT_RUNNING("300003","活动即将开始"),
export const ACTIVITY_NOT_RUNNING = 'ACTIVITY_NOT_RUNNING';
//YOU_HAVE_WIN_ALL_PRIZE("300005","这是你已经拆过的礼包哦~"),
export const YOU_HAVE_WIN_ALL_PRIZE = 'YOU_HAVE_WIN_ALL_PRIZE';
//ACTIVITY_HAS_STOP("300006","活动已结束"),
export const ACTIVITY_HAS_STOP = 'ACTIVITY_HAS_STOP';
//ACTIVITY_HAS_SUSPEND("300007","该活动暂停了"),
export const ACTIVITY_HAS_SUSPEND = 'ACTIVITY_HAS_SUSPEND';
//HIT_NOTHING("300008","很遗憾，你没有抽中礼包~"),
export const HIT_NOTHING = 'HIT_NOTHING';
//NOT_ENOUGH_DRAW_COUNT("300009","抽奖次数不足"),
export const NOT_ENOUGH_DRAW_COUNT = 'NOT_ENOUGH_DRAW_COUNT';
//NO_PERMISSION("300010","没有抽奖资格"),
export const NO_PERMISSION = 'NO_PERMISSION';
//HIT_PRIZE_AND_DRAW_COUNT_PASS_MAX("300011","已领取，赶快使用吧"),
export const HIT_PRIZE_AND_DRAW_COUNT_PASS_MAX = 'HIT_PRIZE_AND_DRAW_COUNT_PASS_MAX';
//AVAILABLE_PRIZE_STOCK_EMPTY("300012","红包已抢完"),
export const AVAILABLE_PRIZE_STOCK_EMPTY = 'AVAILABLE_PRIZE_STOCK_EMPTY';
//NOT_ENOUGH_PRIZE("300013", "很遗憾，你没有抽中礼包~"),
 
export const NOT_ENOUGH_PRIZE = 'NOT_ENOUGH_PRIZE';
//ACTIVITY_QUOTA_EMPTY("300014", "抽奖的额度用完了"),
export const ACTIVITY_QUOTA_EMPTY = 'ACTIVITY_QUOTA_EMPTY';
//Share_For_One_More("300015", "分享再抽一次"),
export const SHARE_FOR_ONE_MORE = 'SHARE_FOR_ONE_MORE';
//NO_PRIZELISE("300020", "还没有抽中任何奖品哦~"),
export const NO_PRIZELISE = 'NO_PRIZELISE';
//NOT_ENOUGH_DAY_DRAW_COUNT("300016", "当天抽奖次数不足"),
export const NOT_ENOUGH_DAY_DRAW_COUNT = 'NOT_ENOUGH_DAY_DRAW_COUNT';
//NOT_OBJECTPRIZE("301001","你并未抽中实物奖"),
export const NOT_OBJECTPRIZE = 'NOT_OBJECTPRIZE';
//LIMIT_KEY_NOT_FOUND("302001", "您的操作过于频繁，请稍后再试！"),
export const LIMIT_KEY_NOT_FOUND = 'LIMIT_KEY_NOT_FOUND';
//LIMIT_BY_RATE("302002", "很遗憾，您没有抽中~"),
export const LIMIT_BY_RATE = 'LIMIT_BY_RATE';
