// @dp/geo/index.js
var Whereami = require("@dp/whereami");
var ua = require("@dp/util-m-ua")();
var fetchJsonp = require('fetch-jsonp');
var Ur = require("@dp/url-rewrite");


var middleIp = {
	cache:false,
	cbQueue:[],
	push:function(cb){
		var self = this;
		self.cbQueue.push(cb);
		self.fetch();
		// 重载
		self.push = function(_cb){
			self.cbQueue.push(_cb);
			if(self.cache)
				self.action();
		}
	},
	fetch:function(){
		var self = this;
		getCityIdByIp(function(data){
			self.cache = data;
			self.lock = true;
			self.action();
		});
	},
	action:function(){
		var self = this;
		var cb;
		while(cb = self.cbQueue.shift()){
			cb(self.cache);
		}
	}
}

function getCityIdByIp(cb){
	fetchJsonp(Ur("//promo.dianping.com/mtlist/getIpInfo.jsonp")).then(function(response){
      try {
        return response.json();
      } catch (e) {	            	
      }
    }).then(function(res){
    	if(res.code == 200){
    		cb({
		    	cityId:res.data
		    });
    	}else{
    		cb({
		    	cityId:""
		    });
    	}
    });
}

function _action(city,cb){
	city = city || {};
	Whereami(function(res){
		var config = {
			lng:res.lng,
	   		lat:res.lat
		}
		if(city.cityId){
	        config.cityId = city.cityId;
	    }else{
	        config.cityId = res.city&&res.city.cityid || "";
	    }
	   	cb(config);
	},function(){
		var timer = false;
		if(city.type){
			cb({
				cityId:city.cityId
			});
		}else{
			middleIp.push(function(data){
				!timer&&cb(data);
				timer = true;
			});
			setTimeout(function(){
				!timer&&cb({
					cityId:""
				});
				timer = true;
			},2500);
		}
	});
}
module.exports =  function(cb,env,options){
	// env 兼容处理
	env = env || ua.source;

	var config = {
		timeout: 3000,
	    city:true,
	    cityType:env,
	    wxAutoConfig:ua.type == "weixin" ? true : false
	}
	options = options || {}
	if(options.useTecentGeo){
		config.useTecentGeo = true;
	}

	// 定位配置
	Whereami.config(config);

	if(ua.type == "mtgroup"){
		require.ensure([], function (require){
            var KNB = require("@dp/knb");
            KNB.ready(function () {
            	//再次给knb打补丁了
            	var _lockGetCity = false;
            	var _timerGetCity = setTimeout(function(){
            		if(!_lockGetCity){
            			_action({},cb);
            			_lockGetCity=true;
            		}
            	},2500);
                KNB.getCity({
                    success:function(city){
                    	if(!_lockGetCity){
                    		if(city && city.cityId == -1){
	                    		_action({},cb);
	                    	}else{
	                    		_action(city,cb)
	                    	}
	                    }

                    	_lockGetCity=true;
                    	clearTimeout(_timerGetCity);
                    },
	      			fail:function(city){
	      				!_lockGetCity&&_action(city,cb);
	      				_lockGetCity=true;
	      				clearTimeout(_timerGetCity);
	      			}
                });
            });
        });
	}else if(ua.type == "dpapp"){
		require.ensure([], function (require) {
            var dpapp = require("@dp/dpapp");
            dpapp.ready(function () {
                dpapp.getCity({
                    success: function(city){_action(city,cb)},
	      			fail:function(city){_action(city,cb)}
                });
            });
        });
	}else{
		_action({},cb);
	}
}
